### PR TITLE
Poly and warmup

### PR DESCRIPTION
**Pull request by harryhan618**
Add a polynomial lr annealing strategy and a lr warmup strategy.
Also, fix an consistency about `LambdaLR` with other lr_scheduler. When epoch=0, Others return base_lr, LambdaLR doesn't. This will cause trouble because in `_LRScheduler`, there is a `self.step()`, and this happens when epoch=0.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -231,6 +231,8 @@ def get_lr(self):
             warnings.warn("To get the last learning rate computed by the scheduler, "
                           "please use `get_last_lr()`.")
 
+        if self.last_epoch == 0:
+            return self.base_lrs
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:235]: This is BC-breaking. Do you have a sample code that gives unexpected results due to this?

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -435,6 +457,45 @@ def _get_closed_form_lr(self):
...
+        if not isinstance(min_lrs, list) and not isinstance(min_lrs, tuple):
+            self.min_lrs = [min_lrs] * len(optimizer.param_groups)
+
+        super().__init__(optimizer, last_epoch=last_epoch)
+
+    def get_lr(self):
+        if self.last_epoch == 0:
+            return self.base_lrs
+
+        base = max(0, 1 - 1 / (self.max_epoch - self.last_epoch - 1))
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:487]: when last_epoch = max_epoch - 1, the denominator is zero?

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -435,6 +457,45 @@ def _get_closed_form_lr(self):
...
+        coeff = base ** coeff
+        return [(lr - min_lr) * coeff + min_lr
+                for group, min_lr in zip(self.optimizer.param_groups, self.min_lrs)]
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:490]: We now make sure that schedulers are chainable were possible, see #26423. This essentially overrides the current value of the scheduler. Is there a way to modify the formula so that the following holds?

> lr(epoch) = lr(epoch-1) * factor

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -314,6 +316,26 @@ def get_lr(self):
             return list(self.base_lrs)
 
 
+class WarmUpLR(LambdaLR):
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:319]: Do you have an example code where you would use this? This looks very close to simply using `MultiplicativeLR` with those formula. Thoughts?

**Review by vincentqb — COMMENTED** [review]
Thanks for working on this! Before you spend more time on this, I'd like to see how we can make those schedulers `PolyLR` fit more nicely in the ecosystem. Thoughts?

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (harryhan618) [reply torch/optim/lr_scheduler.py:235]:
``` python
import torch
import torch.nn as nn
import torch.optim as optim

net = nn.Linear(2, 3)
optimizer = optim.SGD(net.parameters(), 0.1)
print("before lr_scheduler init:", [group['lr'] for group in optimizer.param_groups])

lr_scheduler = optim.lr_scheduler.LambdaLR(optimizer, lambda epoch:0.5)
print("after lr_scheduler init:", [group['lr'] for group in optimizer.param_groups])
```
``` bash
before lr_scheduler init: [0.1]
after lr_scheduler init: [0.05]
```

learning rate changes immediately after lr_scheduler initialized. I'm not sure this is the desired behavior.

File: torch/optim/lr_scheduler.py (hunk 3 above)
Comment (harryhan618) [reply torch/optim/lr_scheduler.py:490]: Oops, `lr` above should be `group['lr']`. fixed.

File: torch/optim/lr_scheduler.py (hunk 2 above)
Comment (harryhan618) [reply torch/optim/lr_scheduler.py:487]: Oops. -1 should be +1. typo, fixed.

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (harryhan618) [reply torch/optim/lr_scheduler.py:319]:
``` python
warmup = WarmUpLR()
for epoch in range(50):
  train(...)
  val(...)
  if epoch < 5:
    warmup.step()
```
In the first 5 epoch, I would use warmup to gradually increase learning rate from low to its origin value. It is normally used in vision task to stable the training process. It would be nice to have such a API. Also, I don't think it is chainable.

**Review by iramazanli — COMMENTED** [review]
Hey ! We currently have already added warmup scheduler. But if you still interested in working on PolyLr i would be happy to collaborate with you.
